src/train.py

```diff
@@ -16,67 +16,248 @@
 import torch.nn as nn
 import torch
 import numpy as np
+from collections import defaultdict
 
 DEVICE = init_device()
 
-# TODO: Refactor these losses...
-def loss_fn(output, target, speech=False):
-    '''
-        Computes the NLL loss between output and target
-    '''
-    PAD_IDX = 0
-    if speech:
-        return F.mse_loss(output, target)
-    else:
-        return F.cross_entropy(output, target, ignore_index=PAD_IDX)
+def text_loss(gold_char, pred_char):
+    return F.cross_entropy(text_pred, char_gold, ignore_index=PAD_IDX)
 
 
-def discriminator_loss(output, target):
-    return F.cross_entropy(output, target)
+def speech_loss(gold_mel, pos_mel, pred_mel, stop_pred):
+
+    # Should be [batch_size x seq_length] for stop 
+    pred_loss = F.mse_loss(pred_mel, gold_mel)
 
+    # TODO: return actual lengths, not just computed off padding
+    # currently, find first nonzero (so pad_idx) in pos_mel, or set to length
+    end_mask_max, end_mask_idx = torch.max((pos_mel == PAD_IDX), dim=1)
+    end_mask_idx[end_mask_max == 0] = pos_mel.shape[1] - 1
+    stop_label = F.one_hot(end_mask_idx, pos_mel.shape[1]).float()
+    stop_loss = F.binary_cross_entropy_with_logits(stop_pred.squeeze(), stop_label)
```

Comment:
nit: need to return the loss?

Reply:
I just returned the sum, following how the loss is computed in Tacotron 2: https://github.com/NVIDIA/tacotron2/blob/master/loss_function.py

However, I think we can definitely add parameters (like lambda_xyz) to weight each of the loss functions in the args.  Any ideas how we might tune these?

Reply:
If it turns out most of the speech output is just an early stop, then we will have to tune this. For now, let's run it and see what happens.